### PR TITLE
feat: [FFM-1861]: Added plan enforcement to Environments page

### PR DESCRIPTION
**Pull request by crystal-fong**
##### Summary:
When accounts exceeds MAU, + Environment button is disabled and tooltip appears 

##### Jira Links:
https://harness.atlassian.net/browse/FFM-1861

##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

**Review by chrisblakely01 — APPROVED** [review]
Looks good - nice work

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  )
+
+  test('it should display plan enforcement tooltip when limits reached', async () => {
+    jest
+      .spyOn(useFeaturesMock, 'useGetFirstDisabledFeature')
+      .mockReturnValue({ featureEnabled: false, disabledFeatureName: FeatureIdentifier.MAUS })
+
+    renderComponent()
+
+    fireEvent.mouseOver(screen.getByText('+ environment'))
```
Comment (knagurski) [inline src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:40]: **Nothing wrong here**, but it's a good practice to use `getByRole` where you can.
```suggestion
    fireEvent.mouseOver(screen.getByRole('button', { name: '+ environment' }))
```

It's a wee bit more verbose, but makes sure we're keeping accessibility in mind. 100% no need to change this, just FYI.

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    jest.spyOn(useFeaturesMock, 'useFeature').mockReturnValue({ enabled: false })
+
+    renderComponent()
+
+    fireEvent.mouseOver(screen.getByText('+ environment'))
+
+    await waitFor(() =>
+      expect(screen.queryByText('common.feature.upgradeRequired.pleaseUpgrade')).not.toBeInTheDocument()
+    )
+    await waitFor(() => expect(screen.getByTestId('create-environment-button')).not.toBeDisabled())
```
Comment (knagurski) [inline src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:58]: Might as well combine these into a single `waitFor`
```suggestion
    await waitFor(() => {
      expect(screen.queryByText('common.feature.upgradeRequired.pleaseUpgrade')).not.toBeInTheDocument()
      expect(screen.getByTestId('create-environment-button')).not.toBeDisabled()
    })
```

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 1 above)
Comment (knagurski) [inline src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:40]: So we're targeting the mouse over based on the text of the button, but you also just added a `data-testid` to the element, can we just use that instead?
```suggestion
    fireEvent.mouseOver(screen.getByTestId('create-environment-button'))
```

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  test('it should hide tooltip and render button when plan enforcement disabled and feature disabled', async () => {
+    jest
+      .spyOn(usePlanEnforcementMock, 'default')
+      .mockReturnValue({ isPlanEnforcementEnabled: false, isFreePlan: false })
+    jest.spyOn(useFeaturesMock, 'useFeature').mockReturnValue({ enabled: false })
+
+    renderComponent()
+
+    fireEvent.mouseOver(screen.getByText('+ environment'))
```
Comment (knagurski) [inline src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:53]: As above, you're referring to the same el with 2 different selectors, might as well use the test id now that it's there.
```suggestion
    fireEvent.mouseOver(screen.getByTestId('create-environment-button'))
```

Or better yet, assign it to a variable and use it in both places.
```suggestion
    const btn = screen.getByTestId('create-environment-button')
    
    fireEvent.mouseOver(btn)
    
    // ...
    
    
    expect(btn).not.toBeDisabled()
```

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  test('it should display plan enforcement tooltip when limits reached', async () => {
+    jest
+      .spyOn(useFeaturesMock, 'useGetFirstDisabledFeature')
+      .mockReturnValue({ featureEnabled: false, disabledFeatureName: FeatureIdentifier.MAUS })
+
+    renderComponent()
+
+    fireEvent.mouseOver(screen.getByText('+ environment'))
+
+    await waitFor(() => expect(screen.getByText('common.feature.upgradeRequired.pleaseUpgrade')).toBeInTheDocument())
```
Comment (knagurski) [inline src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:42]: Do we also want to assert that the button is disabled?

**Review by knagurski — COMMENTED** [review]
Looks great. Just a few suggestions to make the tests a bit more betterer

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 1 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:40]: so would `getByTestId` be even better than `getByRole`?

Comment (chrisblakely01) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:40]: react-testing-library recommends to "test in the way your software is used". In this case the `getByText` more resembles what the user would look for (ie, the button with `+ environment)

Naturally this isn't always feasible thats why getByTestId exists

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 4 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:42]: hm, the button isn't seem to be actually disabled, the prop is just `disabled=""`

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 2 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: done

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 1 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: good to know, thanks!

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 3 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: done

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx (hunk 1 above)
Comment (knagurski) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: `getByRole` is what you really want. The user isn't looking to click on _"+ environment"_, they're looking to click on a **button** that says _"+ environment"_.

Comment (knagurski) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: Personally I'd rip out `getByTestId` and setting the `data-testid` in the component and just use `getByRole`.

Comment (knagurski) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: I'd say the other way around. `getByRole` most closely matches what the user is doing. `getByTestId` is kinda a cheat.

Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx]: updated to getByRole

**Review by chrisblakely01 — DISMISSED** [review]

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      .mockReturnValue({ featureEnabled: false, disabledFeatureName: FeatureIdentifier.MAUS })
+
+    renderComponent()
+
+    const createEnvironmentButton = screen.getByRole('button', { name: '+ environment' })
+    fireEvent.mouseOver(createEnvironmentButton)
+
+    await waitFor(() => {
+      expect(screen.getByText('common.feature.upgradeRequired.pleaseUpgrade')).toBeInTheDocument()
+      expect(createEnvironmentButton.closest('a')).toHaveClass('bp3-disabled')
```
[the remark this answers is not in the document]
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:45]: image

[the remark this answers is not in the document]
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:45]: had to dig deep to find it

[the remark this answers is not in the document]
Comment (knagurski) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:45]: Very weird...
Screenshot 2022-01-19 at 19 04 08

**Review by chrisblakely01 — APPROVED** [review]

**Review by knagurski — APPROVED** [review]
Really solid work, keep it up!

File: src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      .mockReturnValue({ featureEnabled: false, disabledFeatureName: FeatureIdentifier.MAUS })
+
+    renderComponent()
+
+    const createEnvironmentButton = screen.getByRole('button', { name: '+ environment' })
+    fireEvent.mouseOver(createEnvironmentButton)
+
+    await waitFor(() => {
+      expect(screen.getByText('common.feature.upgradeRequired.pleaseUpgrade')).toBeInTheDocument()
+      expect(createEnvironmentButton).toHaveClass('bp3-disabled')
```
[the remark this answers is not in the document]
Comment (crystal-fong) [reply src/modules/75-cf/components/CreateEnvironmentDialog/__tests__/EnvironmentDialog.test.tsx:45]: will remember for next time, thanks!

**Pull request merged**